### ISSUE TITLE
Invalid free in i2o_SCT

### ISSUE THREAD

**Issue by guidovranken** [openssl/openssl#2322]
In ```i2o_SCT()```

```c
223     if (*out != NULL) {
224         p = *out;
225         *out += len;
226     } else {
227         p = OPENSSL_malloc(len);
228         if (p == NULL) {
229             CTerr(CT_F_I2O_SCT, ERR_R_MALLOC_FAILURE);
230             goto err;
231         }
232         *out = p;
233     }
234 
235     if (sct->version == SCT_VERSION_V1) {
236         *p++ = sct->version;
237         memcpy(p, sct->log_id, CT_V1_HASHLEN);
238         p += CT_V1_HASHLEN;
239         l2n8(sct->timestamp, p);
240         s2n(sct->ext_len, p);
241         if (sct->ext_len > 0) {
242             memcpy(p, sct->ext, sct->ext_len);
243             p += sct->ext_len;
244         }
245         if (i2o_SCT_signature(sct, &p) <= 0)
246             goto err;
247     } else {
248         memcpy(p, sct->sct, len);
249     }
250 
251     return len;
252 err:
253     OPENSSL_free(p);
254     return -1;
```

So ```p``` is either allocated, or a function parameter  (```out```).

If version is ```SCT_VERSION_V1``` then, ```p``` is incremented a bit, and ```i2o_SCT_signature``` is called.

If ```i2o_SCT_signature``` fails it jumps to ```err``` and frees ```p```, but ```p``` isn't the original allocated memory address anymore = invalid free. The function also assumes that if ```out``` is supplied, it can be freed, but this is not (always) the case, see calls to ```i2o_SCT``` in ```i2o_SCT_LIST```.

In this particular case,```i2o_SCT_signature``` can only fail if a malloc() fails, so hard to exploit, although I believe this code can be reached remotely. Wasn't sure if I should report this privately but because it depends entirely on a failed malloc I decided to report it here.

PS I haven't tested this, I only read the code, maybe I'm mistaken.

**Comment by guidovranken** [comment]
btw, ```i2o_SCT_signature``` has a similar way of allocating and freeing upon error, but currently not a triggerable invalid free unless you ever add more goto's.

**Comment by levitte** [comment]
Fix in #2323, please try it out
